### PR TITLE
PWA regularly check for service worker updates and automaticially reload the page

### PR DESCRIPTION
**Pull request by jinzo**
Hi,

I was working on the https://github.com/traccar/traccar-web/issues/1169 issue and when deploying to staging I noticed that my PWA (on phone or open in the browser) does not actually update automatically like I would presume with the ```registerType: 'autoUpdate'``` config option set. 
After a bit of digging I actually found two things missing:
1. The ServiceWorker itself did not check for an update when running. So I checked the documentation of VitePWA and adapted the code found there: https://vite-pwa-org.netlify.app/guide/periodic-sw-updates
2. The actual webpage reload (after the ServiceWorker autoUpdates with step 1.) was not applied, therefore I set the immediate parameter to true as seen in the warning here: https://vite-pwa-org.netlify.app/guide/auto-update.html#automatic-reload

I tested this and now it behaves like I would expect it to behave. The Traccar PWA now checks every hour, if there is a new version and if found updates the ServiceWorker and reloads the webpage automaticially. 

On a somewhat related note, just to be sure, so you do prefer automatic reload instead of the update-popup-and-user-clicks-for-reload?

### CODE REVIEW

**Comment by tananaev** [comment]
I thought they have a standard way to detect it automatically? And have a prompt option. How is `registerType: 'autoUpdate'` supposed to work?

**Comment by jinzo** [comment]
I was under that impression too, but apparently not - it did update itself/check (with autoUpdate option) on a new page load (either open a new tab or reload the current page). And that's how I understand this option alone works (it silently updates the ServiceWorker, not the webapp) on new page load (as this is actually not guaranteed with ServiceWorkers, there could be old ServiceWorkers open and somewhere in another tab and they would not get updated if a new tab loads a new version of a ServiceWorker). More info on updating ServiceWorkers: https://web.dev/articles/service-worker-lifecycle#updates
Yes there is a prompt option, if you change registerType to 'prompt' instead of 'autoUpdate' (or omit it, 'prompt' is default) then it won't autoUpdate the ServiceWorker, but it will do it once the user clicks on the button (so it will update the ServiceWorker AND the pwa/web). 
But you still have to check every X ms, if you don't want to wait for (re)load of the webpage (which I'm not even sure how the customer can do if it 'installs' the PWA on their phone for example). 
I personally would prefer the prompt option, so the PWA would just check every X time if there is an update, and use the VitePWA prompt to let the user update if there is one.

**Comment by tananaev** [comment]
It seems like the browser should automatically check for updates:

https://web.dev/articles/service-worker-lifecycle#updates

Maybe I misunderstand something?

**Comment by jinzo** [comment]
The browser is checking for an update in this cases:

- A navigation to an in-scope page.
- A functional events such as push and sync, unless there's been an update check within the previous 24 hours.
- Calling .register() only if the service worker URL has changed. However, you should [avoid changing the worker URL](https://web.dev/articles/service-worker-lifecycle#avoid-url-change).

The last two are not applicable, as we are not pushing or syncing anything to the ServiceWorker and we are not chaging the ServiceWorker URL.
Which means that only the first check is applicable, that happens when a user opens the webpage (either a new load, or an reload of the existing page). So in Traccars React App this means, that if a user has just one tab with it open, and uses it normally (and doesen't close the tab itself) he won't get the updated serviceworker/code until he opens a new one (or reloads the webpage). As a React App doesen't really navigate to a new page and therefore trigger a update. And in the case of a "Installed" PWA on a phone, this probably means force closing the "app" and opening it again.

**Comment by tananaev** [comment]
Why do we need the `immediate: true`? Would having prompt, as we discussed, fix it?

**Comment by jinzo** [comment]
Because of the warning here: https://vite-pwa-org.netlify.app/guide/auto-update.html#automatic-reload
Quote from the yellow WARNING "Automatic reload is not automatic page reload, you will need to use the following code in your application entry point if you want automatic page reload:" My note: Automatic reload == 'autoUpdate'

Yes, having a prompt removes the need for ```immediate: true```

**Comment by tananaev** [comment]
- OK, let's change to prompt
- Let's make auto-update configurable
- Let's keep this copied code in a separate file and provide a link to the original

**Comment by jinzo** [comment]
Ok, just to be clear, so user configurable as a server setting? And do we make it just on/off or can the update check interval be configured too (eg. then I would still go with one setting, but if it's 0 then don't check, if it's more than 0 then use that as an update check interval)?

**Comment by tananaev** [comment]
We can make the interval configurable, if that's easy to implement. And zero means no updates. I think we should include it in the server attributes.

**Comment by tananaev** [comment]
I think if it works well, let's make it as a default, so some reasonable default value.

**Comment by jinzo** [comment]
So this is just a work in progress for now - but have to finish for today and comited to get some preliminary feedback.

I moved everything to a React component and adapted it to use plain Snackbar like other similar functionality in Traccar. While tweaking this I tested it plenty, and it works very well. What remains is figuring out why "useAttributePreference" is erroring out on me and do basic cleanup around that. I'm thinking 1h as the default value (as this is also the 'default' in Vite PWA documentation).

P.S: There is a big ol red warning that switching between 'autoUpdate' and 'prompt' while in production can be a pain, but from what I can see 'autoUpdate' did not make it into 5.9 so this won't be a problem: https://vite-pwa-org.netlify.app/guide/auto-update.html

**Comment by jinzo** [comment]
After stepping back from the monitor for a few minutes it dawned on me, that as the system is not initialized, I cannot access the user attributes. Moved it to a better location and it is ready for a review.

File: modern/src/ReloadPrompt.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { Snackbar, Button } from '@mui/material';
+import React from 'react'
+import { useTranslation } from './common/components/LocalizationProvider';
+import { useAttributePreference } from './common/util/preferences';
+import { useRegisterSW } from 'virtual:pwa-register/react'
+
+// Based on https://vite-pwa-org.netlify.app/frameworks/react.html
+function ReloadPrompt() {
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:8]: We need a better name because it's not just prompt. It actually checks.

File: modern/src/ReloadPrompt.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const t = useTranslation();
+
+  const serviceWorkerUpdateInterval = useAttributePreference('serviceWorkerUpdateInterval', 3600000);
+
+  const {
+    offlineReady: [offlineReady, setOfflineReady],
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:14]: Why do we need this?

File: modern/src/ReloadPrompt.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    needRefresh: [needRefresh, setNeedRefresh],
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:15]: And `setNeedRefresh`?

File: modern/src/ReloadPrompt.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import React from 'react'
+import { useTranslation } from './common/components/LocalizationProvider';
+import { useAttributePreference } from './common/util/preferences';
+import { useRegisterSW } from 'virtual:pwa-register/react'
+
+// Based on https://vite-pwa-org.netlify.app/frameworks/react.html
+function ReloadPrompt() {
+  const t = useTranslation();
+
+  const serviceWorkerUpdateInterval = useAttributePreference('serviceWorkerUpdateInterval', 3600000);
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:11]: This should be server only. Also we usually have values in seconds or milliseconds. This seems to be in hours?

File: modern/src/ReloadPrompt.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+          cache: 'no-store',
+          headers: {
+            'cache': 'no-store',
+            'cache-control': 'no-cache',
+          },
+        });
+  
+        if (resp?.status === 200) {
+          await r.update();
+        }
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:38]: Can we give proper names to variables please.

File: modern/src/ReloadPrompt.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      action={(
+        <Button onClick={() => updateServiceWorker(true)}>
+          {t('sharedReload')}
+        </Button>
+      )}
+      />
+  );
+}
+
+export default ReloadPrompt;
```
Comment (tananaev) [inline modern/src/ReloadPrompt.tsx:56]: Missing new line

File: modern/src/resources/l10n/en.json [hunk 7]
```diff
@@ -99,6 +99,8 @@
     "sharedImport": "Import",
     "sharedColumns": "Columns",
     "sharedDropzoneText": "Drag and drop a file here or click",
+    "sharedReload": "Reload",
+    "updateAvailable": "New update available, click on the reload button to update.",
```
Comment (tananaev) [inline modern/src/resources/l10n/en.json:103]: This should have a prefix. Maybe it should all be under "settings" prefix?

Comment (tananaev) [reply modern/src/resources/l10n/en.json:103]: Also I think the wording can be simplified. If there's a clear reload button, there's no need to tell that you have to reload.

Comment (tananaev) [reply modern/src/resources/l10n/en.json:103]: Please include a screenshot.

File: modern/src/ReloadPrompt.tsx (hunk 1 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx:8]: UpdateCheckPrompt ? or even UpdateCheckAndPrompt?

File: modern/src/ReloadPrompt.tsx (hunk 2 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx:14]: Sorry this remained from the original snippet, will remove it.

File: modern/src/ReloadPrompt.tsx (hunk 3 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx:15]: Sorry this remained from the original snippet, will remove it.

File: modern/src/ReloadPrompt.tsx (hunk 4 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx:11]: I'm still not 100% sure I understand different settings and possibilities in Traccar. Are we talking about here checking only ```state.session.server.attributes.serviceWorkerUpdateInterval``` right? or are we talking about traccar.conf on the server side?
Also value is in milliseconds 60 * 60 * 1000 = 3600000

File: modern/src/ReloadPrompt.tsx (hunk 5 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx:38]: I mostly left this part with minimum changes from the original snippet, but will rename.

File: modern/src/resources/l10n/en.json (hunk 7 above)
Comment (jinzo) [reply modern/src/resources/l10n/en.json:103]: Simplified the wording, see attached screenshot of a mobile view.
Screenshot from 2023-11-06 09-47-12

File: modern/src/ReloadPrompt.tsx (hunk 1 above)
Comment (jinzo) [reply modern/src/ReloadPrompt.tsx]: Went with UpdateCheckPrompt for now, after looking around UpdateController could be used too. What do you think?

File: modern/src/resources/l10n/en.json (hunk 7 above)
Comment (jinzo) [reply modern/src/resources/l10n/en.json]: There is also a possibility of outlined button
Screenshot from 2023-11-06 13-47-12

File: modern/src/UpdateCheckPrompt.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  });
+
+  return (
+    <Snackbar 
+      open={needRefresh} 
+      message={t('settingsUpdateAvailable')}
+      action={(
+        <Button onClick={() => updateServiceWorker(true)}>
+          {t('settingsReload')}
+        </Button>
```
Comment (tananaev) [inline modern/src/UpdateCheckPrompt.tsx:51]: Can we just use an icon button? Like this one:

https://mui.com/material-ui/material-icons/?query=ref&selected=Refresh

It looks better and also we can remove one of the strings.

File: modern/src/ReloadPrompt.tsx (hunk 4 above)
Comment (tananaev) [reply modern/src/ReloadPrompt.tsx]: We're talking about `state.session.server.attributes.serviceWorkerUpdateInterval`.

File: modern/vite-env.d.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+/// <reference types="vite-plugin-pwa/react" />
```
[the remark this answers is not in the document]
Comment (tananaev) [reply modern/vite-env.d.ts:1]: What is it complaining about? I suggest we remove this. We're not planning to make any changes to that file anyway.

File: modern/src/UpdateCheckPrompt.tsx (hunk 8 above)
Comment (jinzo) [reply modern/src/UpdateCheckPrompt.tsx]: Good idea, did not even think about this.
image

File: modern/vite-env.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jinzo) [reply modern/vite-env.d.ts]: image
Removed it.

File: modern/src/common/util/preferences.js [hunk 10]
```diff
@@ -39,3 +39,10 @@ export const useAttributePreference = (key, defaultValue) => useSelector((state)
   }
   return defaultValue;
 });
+
+export const useServerAttributePreference = (key, defaultValue) => useSelector((state) => {
```
Comment (tananaev) [inline modern/src/common/util/preferences.js:43]: It doesn't make sense to add this function. You can simply write:
```js
state.session.server.attributes.key || defaultValue
```

Comment (jinzo) [reply modern/src/common/util/preferences.js]: Thank you, did not know that.

**Comment by tananaev** [comment]
Merged, thank you.

**Comment by jinzo** [comment]
Thank you very much for your patience.

**Pull request merged**